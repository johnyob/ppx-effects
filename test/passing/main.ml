exception%effect E : string

let comp () =
  print_string "0 ";
  print_string (perform E);
  print_string "3 "
;;

let () =
  try comp () with
  | [%effect? E, k] ->
    print_string "1 ";
    continue k "2 ";
    print_string "4 "
;;

let () =
  match comp () with
  | e -> e
  | [%effect? E, k] ->
    print_string "1 ";
    continue k "2 ";
    print_string "4 "
;;
