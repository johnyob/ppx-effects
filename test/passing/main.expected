open! Ppx_effects_runtime.Import
type _ Ppx_effects_runtime.t +=  
  | E: string Ppx_effects_runtime.t 
let comp () = print_string "0 "; print_string (perform E); print_string "3 "
let () =
  Ppx_effects_runtime.Deep.try_with comp ()
    {
      Ppx_effects_runtime.Deep.effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function
              | E ->
                  Some
                    ((fun k ->
                        print_string "1 "; continue k "2 "; print_string "4 "))
              | _ -> None : continue_input Ppx_effects_runtime.t ->
                              ((continue_input, _)
                                 Ppx_effects_runtime.Deep.continuation -> 
                                 _)
                                option) in
         effc)
    }
let () =
  Ppx_effects_runtime.Deep.match_with comp ()
    {
      Ppx_effects_runtime.Deep.retc = (function | e -> e);
      exnc = Ppx_effects_runtime.raise;
      effc =
        (let effc (type continue_input) =
           (let open! Ppx_effects_runtime.Deep.Open_on_rhs in
              function
              | E ->
                  Some
                    ((fun k ->
                        print_string "1 "; continue k "2 "; print_string "4 "))
              | _ -> None : continue_input Ppx_effects_runtime.t ->
                              ((continue_input, _)
                                 Ppx_effects_runtime.Deep.continuation -> 
                                 _)
                                option) in
         effc)
    }
